
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="SnpEff and SnpSift">
      
      
        <meta name="author" content="Pablo Cingolani">
      
      
      
        <link rel="prev" href="../inputoutput/">
      
      
        <link rel="next" href="../additionalann/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.19">
    
    
      
        <title>Cancer samples - SnpEff & SnpSift</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.eebd395e.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="default" data-md-color-accent="indigo">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cancer-samples" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-header__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SnpEff & SnpSift
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Cancer samples
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../introduction/" class="md-tabs__link md-tabs__link--active">
        SnpEff
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../snpsift/introduction/" class="md-tabs__link">
        SnpSift
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../download/" class="md-tabs__link">
        Download & Install
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../examples/" class="md-tabs__link">
        Examples
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../about/" class="md-tabs__link">
        About
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-nav__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    SnpEff & SnpSift
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
      
      
      
        <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Usage examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../license/" class="md-nav__link">
        License
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          SnpEff
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          SnpEff
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../running/" class="md-nav__link">
        Running SnpEff
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../commandline/" class="md-nav__link">
        Commands &amp; command line options
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../inputoutput/" class="md-nav__link">
        Input &amp; output files
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Cancer samples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Cancer samples
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#running-in-cancer-analysis-mode" class="md-nav__link">
    Running in cancer analysis mode
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#representing-cancer-data" class="md-nav__link">
    Representing cancer data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#defining-cancer-samples" class="md-nav__link">
    Defining cancer samples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interpreting-cancer-annotations" class="md-nav__link">
    Interpreting Cancer annotations
  </a>
  
    <nav class="md-nav" aria-label="Interpreting Cancer annotations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#cancer-annotations-using-eff-field" class="md-nav__link">
    Cancer annotations using 'EFF' field:
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../additionalann/" class="md-nav__link">
        Additional annotations
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../outputsummary/" class="md-nav__link">
        Output summary files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        Troubleshooting
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db/" class="md-nav__link">
        Building databases
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db_gff_gtf/" class="md-nav__link">
        Build databses. GTF / GFF details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_reg/" class="md-nav__link">
        Building databases. Regulatory and Non-coding
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_pdb/" class="md-nav__link">
        Building databases. PDB and AlphaFold
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../integration/" class="md-nav__link">
        Integration. GATK and Galaxy
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../commands/" class="md-nav__link">
        Commands and utilities
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../human_genomes/" class="md-nav__link">
        Human Genomes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        Frequently Asked Questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          SnpSift
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          SnpSift
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/filter/" class="md-nav__link">
        SnpSift Filter
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/annotate/" class="md-nav__link">
        SnpSift Annotate
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/casecontrol/" class="md-nav__link">
        SnpSift CaseControl
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervals/" class="md-nav__link">
        SnpSift Intervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervalsindex/" class="md-nav__link">
        SnpSift Intervals Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/join/" class="md-nav__link">
        SnpSift Join
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rmrefgen/" class="md-nav__link">
        SnpSift RmRefGen
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/tstv/" class="md-nav__link">
        SnpSift TsTv
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/extractfields/" class="md-nav__link">
        SnpSift Extract Fields
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/varianttype/" class="md-nav__link">
        SnpSift Variant Type
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gwascatalog/" class="md-nav__link">
        SnpSift GWAS Catalog
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/dbnsfp/" class="md-nav__link">
        SnpSift dbNSFP
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/split/" class="md-nav__link">
        SnpSift Split
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/phastcons/" class="md-nav__link">
        SnpSift PhastCons
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/concordance/" class="md-nav__link">
        SnpSift Concordance
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/private/" class="md-nav__link">
        SnpSift Private
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcf2ped/" class="md-nav__link">
        SnpSift Vcf2Tped
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intersect/" class="md-nav__link">
        SnpSift Intersect
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rminfo/" class="md-nav__link">
        SnpSift RmInfo
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/genesets/" class="md-nav__link">
        SnpSift GeneSets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gt/" class="md-nav__link">
        SnpSift GT
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcfcheck/" class="md-nav__link">
        SnpSift VcfCheck
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/faq/" class="md-nav__link">
        Frequently asked questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          Download & Install
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Download & Install
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../download/" class="md-nav__link">
        Download and install
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          Examples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Examples
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Examples
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          About
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../help/" class="md-nav__link">
        Help and Bugs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../xiangyi_lu_donate/" class="md-nav__link">
        In Memory
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="cancer-samples">Cancer samples</h1>
<p>Here we describe details about annotating cancer samples.</p>
<p><strong> Single multi-sample VCF vs. multiple VCF files. </strong></p>
<p>It is common practice, to have all samples in a single "multi-sample VCF file" (having two or more separate VCF files is highly discouraged).
This is also the "gold standard" in cancer analysis standard, so all samples (both somatic and germline) should be in one VCF file.</p>
<p>SnpEff requires that you follow gold standard practices, thus requires a single multi-sample VCF (it is not possible to run cancer analysis using multiple VCF files).</p>
<h3 id="running-in-cancer-analysis-mode">Running in cancer analysis mode</h3>
<p>Using the <code>-cancer</code> command line option, you can compare somatic vs germline samples.</p>
<p>So an example command line would be:</p>
<div class="codehilite"><pre><span></span><code>$<span class="w"> </span>java<span class="w"> </span>-Xmx8g<span class="w"> </span>-jar<span class="w"> </span>snpEff.jar<span class="w"> </span>-v<span class="w"> </span>-cancer<span class="w"> </span>GRCh37.75<span class="w"> </span>cancer.vcf<span class="w"> </span>&gt;<span class="w"> </span>cancer.ann.vcf
</code></pre></div>

<h3 id="representing-cancer-data">Representing cancer data</h3>
<p>In a typical cancer sequencing experiment, we want to measure and annotate differences between germline (healthy) and somatic (cancer) tissue samples from the same patient.
The complication is that germline is not always the same as the reference genome, so a typical annotation does not work.</p>
<p>For instance, let's assume that at a given genomic position (e.g. chr1:69091), reference genome is 'A', germline is 'C' and somatic is 'G'.
This should be represented in a VCF file as:
<div class="highlight"><pre><span></span><code>#CHROM  POS    ID  REF  ALT    QUAL  FILTER  INFO  FORMAT  Patient_01_Germline  Patient_01_Somatic
1       69091  .   A    C,G    .     PASS    AC=1  GT      1/0                  2/1
</code></pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Some people tend to represent this by changing REF base 'A' using germline 'C'.
This is a mistake, REF must always represent the reference genome, not one of your samples.</p>
</div>
<p>Under normal conditions, SnpEff would provide the effects of changes "A -&gt; C" and "A -&gt; G".
But in case of cancer samples, we are actually interested in the difference between somatic and germline, so we'd like to calculate the effect of a "C -&gt; G" mutation.
Calculating this effect is not trivial, since we have to build a new "reference" by calculating the effect of the first mutation ("A -&gt; C") and then calculate the effect of the second one ("C -&gt; G") on our "new reference".</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>In order to activate cancer analysis, you must use <code>-cancer</code> command line option.</p>
</div>
<h3 id="defining-cancer-samples">Defining cancer samples</h3>
<p>As we already mentioned, cancer data is represented in a VCF file using multiple ALTs (REF field always is reference genome).
In order to specify which samples are somatic and which ones are germline, there are two options:</p>
<ul>
<li>Use a TXT file using <code>-cancerSamples</code> command line option.</li>
<li>Use the PEDIGREE meta information in your VCF file's header. This is the default, but some people might find hard to edit / change information in VCF file's headers.</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you do not provide either <code>PEDIGREE</code> meta information or a TXT samples file, SnpEff will not know which somatic samples derive from which germline samples.
Thus it will be unable to perform cancer effect analysis.</p>
</div>
<p><strong>TXT file</strong></p>
<p>This is quite easy.
All you have to do is to create a tab-separated TXT file having two columns: the first column has the germline sample names and the second column has the somatic sample names.
Make sure that sample names match exactly the ones in the VCF file.</p>
<p>E.g.: Create a TXT file named 'samples_cancer.txt'
<div class="highlight"><pre><span></span><code>Patient_01_Germline Patient_01_Somatic
Patient_02_Germline Patient_02_Somatic
Patient_03_Germline Patient_03_Somatic
Patient_04_Germline Patient_04_Somatic
</code></pre></div>
Then you have to specify this TXT file when invoking SnpEff, using the <code>-cancerSamples</code> command line option.</p>
<p>E.g. In our example, the file name is 'samples_cancer.txt', so the command line would look like this:
<div class="highlight"><pre><span></span><code>$ cat examples/samples_cancer_one.txt
Patient_01_Germline    Patient_01_Somatic

$ java -Xmx8g -jar snpEff.jar -v \
                -cancer \
                -cancerSamples examples/samples_cancer_one.txt \
                GRCh37.75 \
                examples/cancer.vcf \
                &gt; cancer.ann.vcf
</code></pre></div>
<strong>VCF header</strong></p>
<p>This is the default method and the main advantage is that you don't have to carry information on a separate TXT file (all the information is within your VCF file).
You have to add the <code>PEDIGREE</code> header with the appropriate information to your VCF file.
Obviously this requires you to edit you VCF file's header.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>How to edit VCF headers is beyond the scope of this manual (we recommend using <code>vcf-annotate</code> from VCFtools). But if you find adding PEDIGREE information to your VCF file difficult, just use the TXT file method described in the previous sub-section.</p>
</div>
<p>E.g.: Pedigree information in a VCF file would look like this:
<div class="highlight"><pre><span></span><code>$ cat examples/cancer_pedigree.vcf
##PEDIGREE=&lt;Derived=Patient_01_Somatic,Original=Patient_01_Germline&gt;
#CHROM  POS ID  REF ALT QUAL    FILTER  INFO        FORMAT  Patient_01_Germline Patient_01_Somatic
1   69091   .   A   C,G .       PASS    AF=0.1122   GT      1/0                 2/1
1   69849   .   G   A,C .       PASS    AF=0.1122   GT      1/0                 2/1
1   69511   .   A   C,G .       PASS    AF=0.3580   GT      1/1                 2/2


$ java -Xmx8g -jar snpEff.jar -v -cancer GRCh37.75 examples/cancer_pedigree.vcf &gt; examples/cancer_pedigree.ann.vcf
</code></pre></div>
Here we say that the sample called <code>Patient_01_Somatic</code> is derived from the sample called <code>Patient_01_Germline</code>.
In this context, this means that cancer sample is derived from the healthy tissue.</p>
<h3 id="interpreting-cancer-annotations">Interpreting Cancer annotations</h3>
<p>Interpretation of <code>ANN</code> field cancer sample relies on 'Allele' sub-field.
Just as a reminder, <code>ANN</code> field has the following format:</p>
<div class="highlight"><pre><span></span><code>ANN = Allele | Annotation | Annotation_Impact | Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | AA.pos / AA.length | Distance | ERRORS_WARNINGS_INFO
</code></pre></div>
<p>The <code>Allele</code> field tells you which effect relates to which genotype.
More importantly, genotype difference between Somatic and Germline.</p>
<p>Example: when there are multiple ALTs (e.g. REF='A' ALT='C,G') and the genotype field says:</p>
<ul>
<li>Allele = "C": it means is the effect related to the first ALT ('C')</li>
<li>Allele = "G" if it's the effect related to the second ALT ('G')</li>
<li>Allele = "G-C" means that this is the effect of having the second ALT as variant while using the first ALT as reference ("C -&gt; G").
It is important that you understand the meaning of the last one, because you'll use it often for your cancer analysis.</li>
</ul>
<p>Example: Sample output for the previously mentioned VCF example would be (the output has been edited for readability reasons)</p>
<p>For the first line we get (edited for readability):
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar -v -cancer -cancerSamples examples/samples_cancer_one.txt GRCh37.75 examples/cancer.vcf &gt; examples/cancer.eff.vcf

1   69091   .   A   C,G .   PASS    AF=0.1122;
                  ANN=G|start_lost|HIGH|OR4F5|ENSG00000186092|transcript|ENST00000335137|protein_coding|1/1|c.1A&gt;G|p.Met1?|1/918|1/918|1/305||
                 ,G-C|start_lost|HIGH|OR4F5|ENSG00000186092|transcript|ENST00000335137|protein_coding|1/1|c.1A&gt;G|p.Leu1?|1/918|1/918|1/305||
                 ,C|initiator_codon_variant|LOW|OR4F5|ENSG00000186092|transcript|ENST00000335137|protein_coding|1/1|c.1A&gt;C|p.Met1?|1/918|1/918|1/305||
               GT  1/0 2/1
</code></pre></div>
What does it mean:</p>
<ol>
<li>In this case, we have two ALTs = 'C' and 'G'.</li>
<li>Germline sample is heterozygous 'C/A' (GT = '1/0')</li>
<li>Somatic tissue is heterozygous 'G/C' (GT = '2/1')</li>
<li>
<p>Change A -&gt; C and A -&gt; G are always calculated by SnpEff (this is the "default mode").</p>
<ul>
<li>
<p>A -&gt; C produces this effect:</p>
<div class="codehilite"><pre><span></span><code><span class="n">C</span><span class="o">|</span><span class="n">initiator_codon_variant</span><span class="o">|</span><span class="n">LOW</span><span class="o">|</span><span class="n">OR4F5</span><span class="o">|</span><span class="n">ENSG00000186092</span><span class="o">|</span><span class="n">transcript</span><span class="o">|</span><span class="n">ENST00000335137</span><span class="o">|</span><span class="n">protein_coding</span><span class="o">|</span><span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="o">|</span><span class="n">c</span><span class="o">.</span><span class="mi">1</span><span class="n">A</span><span class="o">&gt;</span><span class="n">C</span><span class="o">|</span><span class="n">p</span><span class="o">.</span><span class="n">Met1</span><span class="err">?</span><span class="o">|</span><span class="mi">1</span><span class="o">/</span><span class="mi">918</span><span class="o">|</span><span class="mi">1</span><span class="o">/</span><span class="mi">918</span><span class="o">|</span><span class="mi">1</span><span class="o">/</span><span class="mi">305</span><span class="o">||</span>
</code></pre></div>

<p>Note that the last field (genotype field) is 'C' indicating this is produced by the first ALT.</p>
</li>
<li>
<p>A -&gt; G produces this effect:</p>
<div class="codehilite"><pre><span></span><code>G|start_lost|HIGH|OR4F5|ENSG00000186092|transcript|ENST00000335137|protein_coding|1/1|c.1A&gt;G|p.Met1?|1/918|1/918|1/305||
</code></pre></div>

<p>Note that the last field (genotype field) is 'G' indicating this is produced by the second ALT.</p>
</li>
</ul>
</li>
<li>
<p>Finally, this is what you were expecting for, the cancer comparisons. Since both germline and somatic are heterozygous (GT are '1/0' and '2/1'), there are 4 possible comparisons to make:</p>
<ul>
<li>G vs C : This is the Somatic vs Germline we are interested in. SnpEff reports this one</li>
<li>G vs A : This compares ALT to REF, so it was already reported in "default mode". SnpEff doesn't report this one again.</li>
<li>C vs C : This is not a variant, since both og them ar '1'. SnpEff skips this one.</li>
<li>C vs A : This compares ALT to REF, so it was already reported in "default mode". SnpEff doesn't report this one again.</li>
</ul>
<p>I know is confusing, but the bottom line is that only the first comparison one makes sense, and is the one SnpEff reports.
So 'C -&gt; G' produces the following effect:</p>
<div class="codehilite"><pre><span></span><code>  G-C|start_lost|HIGH|OR4F5|ENSG00000186092|transcript|ENST00000335137|protein_coding|1/1|c.1A&gt;G|p.Leu1?|1/918|1/918|1/305||
</code></pre></div>

<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Notice the genotype field is "G-C" meaning the we produce a new reference on the fly using ALT 1 ('C') and then used ALT 2 ('G') as the variant. So we compare 'G' (ALT) to 'C' (REF).</p>
</div>
</li>
</ol>
<h4 id="cancer-annotations-using-eff-field">Cancer annotations using 'EFF' field:</h4>
<p>Interpretation of <code>EFF</code> field cancer sample relies on 'Genotype' sub-field.
Just as a reminder, <code>EFF</code> field has the following format:
<div class="highlight"><pre><span></span><code>EFF = Effect ( Effect_Impact | Functional_Class | Codon_Change | Amino_Acid_Change| Amino_Acid_Length | Gene_Name | Transcript_BioType | Gene_Coding | Transcript_ID | Exon_Rank  | Genotype_Number [ | ERRORS | WARNINGS ] )
</code></pre></div></p>
<p>For the previous example, we get (edited for readability):
<div class="highlight"><pre><span></span><code>$ java -Xmx8g -jar snpEff.jar -v -classic -cancer -cancerSamples examples/samples_cancer_one.txt GRCh37.75 examples/cancer.vcf &gt; examples/cancer.eff.vcf

1   69091   .   A   C,G .   PASS    AC=1;
                     EFF=START_LOST(HIGH|MISSENSE|Atg/Gtg|M1V|305|OR4F5|protein_coding|CODING|ENST00000335137|1|G)
                        ,START_LOST(HIGH|MISSENSE|Ctg/Gtg|L1V|305|OR4F5|protein_coding|CODING|ENST00000335137|1|G-C)
                        ,NON_SYNONYMOUS_START(LOW|MISSENSE|Atg/Ctg|M1L|305|OR4F5|protein_coding|CODING|ENST00000335137|1|C)
</code></pre></div></p>
<p>The <code>GenotypeNum</code> field tells you which effect relates to which genotype.
More importantly, genotype difference between Somatic and Germline.</p>
<p>Example: when there are multiple ALTs (e.g. REF='A' ALT='C,G') and the genotype field says:</p>
<ul>
<li>GenotypeNum = "1": it means is the effect related to the first ALT ('C')</li>
<li>GenotypeNum = "2" if it's the effect related to the second ALT ('G')</li>
<li>GenotypeNum = "2-1" means that this is the effect of having the second ALT as variant while using the first ALT as reference ("C -&gt; G").</li>
</ul>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Pablo Cingolani
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="http://www.linkedin.com/in/pablocingolani" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/pcingola/SnpEff" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "search.highlight", "search.suggest", "toc.integrate", "content.tabs.link", "content.code.annotate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.220ee61c.min.js"></script>
      
    
  </body>
</html>