
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="SnpEff and SnpSift">
      
      
        <meta name="author" content="Pablo Cingolani">
      
      
      
        <link rel="prev" href="../additionalann/">
      
      
        <link rel="next" href="../troubleshooting/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.19">
    
    
      
        <title>Output summary files - SnpEff & SnpSift</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.eebd395e.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="default" data-md-color-accent="indigo">
  
    
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#output-summary-files" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-header__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SnpEff & SnpSift
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output summary files
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  
    
  


  
  
  
    <li class="md-tabs__item">
      <a href="../introduction/" class="md-tabs__link md-tabs__link--active">
        SnpEff
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../snpsift/introduction/" class="md-tabs__link">
        SnpSift
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../download/" class="md-tabs__link">
        Download & Install
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../examples/" class="md-tabs__link">
        Examples
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../../about/" class="md-tabs__link">
        About
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="SnpEff &amp; SnpSift" class="md-nav__button md-logo" aria-label="SnpEff & SnpSift" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    SnpEff & SnpSift
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/pcingola/SnpEff" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    SnpEff
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
      
      
      
        <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Usage examples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../license/" class="md-nav__link">
        License
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
      
      
      
        <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
          SnpEff
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          SnpEff
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../running/" class="md-nav__link">
        Running SnpEff
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../commandline/" class="md-nav__link">
        Commands &amp; command line options
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../inputoutput/" class="md-nav__link">
        Input &amp; output files
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../cansersamples/" class="md-nav__link">
        Cancer samples
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../additionalann/" class="md-nav__link">
        Additional annotations
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Output summary files
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Output summary files
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#summary-file-options" class="md-nav__link">
    Summary file options
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-report" class="md-nav__link">
    Summary report
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#main-summary-table" class="md-nav__link">
    Main summary table
  </a>
  
    <nav class="md-nav" aria-label="Main summary table">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#variants-by-type" class="md-nav__link">
    Variants by type
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#histograms" class="md-nav__link">
    Histograms
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotations-region" class="md-nav__link">
    Annotations &amp; Region
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-statistics" class="md-nav__link">
    Gene statistics
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../troubleshooting/" class="md-nav__link">
        Troubleshooting
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db/" class="md-nav__link">
        Building databases
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_db_gff_gtf/" class="md-nav__link">
        Build databses. GTF / GFF details
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_reg/" class="md-nav__link">
        Building databases. Regulatory and Non-coding
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../build_pdb/" class="md-nav__link">
        Building databases. PDB and AlphaFold
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../integration/" class="md-nav__link">
        Integration. GATK and Galaxy
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../commands/" class="md-nav__link">
        Commands and utilities
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../human_genomes/" class="md-nav__link">
        Human Genomes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../faq/" class="md-nav__link">
        Frequently Asked Questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          SnpSift
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          SnpSift
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/introduction/" class="md-nav__link">
        Introduction
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/filter/" class="md-nav__link">
        SnpSift Filter
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/annotate/" class="md-nav__link">
        SnpSift Annotate
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/casecontrol/" class="md-nav__link">
        SnpSift CaseControl
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervals/" class="md-nav__link">
        SnpSift Intervals
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intervalsindex/" class="md-nav__link">
        SnpSift Intervals Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/join/" class="md-nav__link">
        SnpSift Join
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rmrefgen/" class="md-nav__link">
        SnpSift RmRefGen
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/tstv/" class="md-nav__link">
        SnpSift TsTv
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/extractfields/" class="md-nav__link">
        SnpSift Extract Fields
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/varianttype/" class="md-nav__link">
        SnpSift Variant Type
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gwascatalog/" class="md-nav__link">
        SnpSift GWAS Catalog
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/dbnsfp/" class="md-nav__link">
        SnpSift dbNSFP
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/split/" class="md-nav__link">
        SnpSift Split
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/phastcons/" class="md-nav__link">
        SnpSift PhastCons
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/concordance/" class="md-nav__link">
        SnpSift Concordance
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/private/" class="md-nav__link">
        SnpSift Private
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcf2ped/" class="md-nav__link">
        SnpSift Vcf2Tped
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/intersect/" class="md-nav__link">
        SnpSift Intersect
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/rminfo/" class="md-nav__link">
        SnpSift RmInfo
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/genesets/" class="md-nav__link">
        SnpSift GeneSets
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/gt/" class="md-nav__link">
        SnpSift GT
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/vcfcheck/" class="md-nav__link">
        SnpSift VcfCheck
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../snpsift/faq/" class="md-nav__link">
        Frequently asked questions
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          Download & Install
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Download & Install
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../download/" class="md-nav__link">
        Download and install
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
      
      
      
        <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
          Examples
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_5">
          <span class="md-nav__icon md-icon"></span>
          Examples
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../examples/" class="md-nav__link">
        Examples
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
      
      
      
        <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
          About
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_6">
          <span class="md-nav__icon md-icon"></span>
          About
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../about/" class="md-nav__link">
        About
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../help/" class="md-nav__link">
        Help and Bugs
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../../xiangyi_lu_donate/" class="md-nav__link">
        In Memory
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="output-summary-files">Output summary files</h1>
<p>SnpEff creates an additional output file showing overall statistics.
This "stats" file is an HTML file which can be opened using a web browser.
It can also be created as a CSV file, for easier parsing or manipulation.
You can find an example of a 'stats' file <a href="../../adds/1kg.html">here</a>.</p>
<h2 id="summary-file-options">Summary file options</h2>
<p>By default, SnpEff performs some statistics and saves them to the <code>snpEff_summary.html</code> on the same directory where snpEff is being executed.
You can see the file, by opening it in your browser. SnpEff also creates a "Genes Statistics" file (<code>snpEff_genes.txt</code>) with gene-level statistics in tabular (tab-separated) format.</p>
<p>There are some command line options related to the statistics:</p>
<ul>
<li><code>-stats &lt;file&gt;</code>: You can change the default name and location of the HTML file. This also changes the name and location of the "genes" (TXT) file.</li>
<li><code>-noStats</code>: Do not calculate statisticsm, or create stats (summary) files</li>
<li><code>-csvStats &lt;file&gt;</code>: Create the statistics file in CSV format with the specified name.</li>
</ul>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The "Genes statistics" file path is by default the directory where SnpEff is executed and called <code>snpEff_genes.txt</code>. If you change the summary file name / path by either using <code>-stats</code> or <code>csvStats</code> command line, the "Genes statistics" file path will be the same directory as the summary file and the file name is the same "base name" plus a <code>".genes.txt"</code>.</p>
</div>
<h2 id="summary-report">Summary report</h2>
<p>The summary report consist of several sections:</p>
<ul>
<li><a href="#main-summary-table">Main summary table</a></li>
<li>Variant rate by chromosome: A table of number of variants and variants rate per chromosome</li>
<li><a href="#variants-by-type">Variants by type</a></li>
<li>Number of variants by impact: A count and percentage of variants by <a href="../inputoutput/#impact-prediction">impact</a></li>
<li>Number of variants by functional class: A count and percentage of variants by <a href="../inputoutput/#functional-class">functional class</a></li>
<li><a href="#annotations--region">Number of variants by annotation</a></li>
<li>Quality histogram: Variant Quality histogram, i.e. a histogram of the VCF's  <code>QUAL</code> field</li>
<li>InDel length histogram: Length of <code>INS</code> and <code>DEL</code> variants</li>
<li>Base variant table: Table of SNV variants changes</li>
<li>Transition vs transversions (ts/tv): Transition vs transversions table, split into "known" and all variants (known variants are the ones with non-empty <code>ID</code> field).</li>
<li>Allele frequency: Allele frequency histogram</li>
<li>Allele Count: Allele count histogram</li>
<li>Codon change table: A table of all codon changes counts</li>
<li>Amino acid change table: A table of all amino acid changes counts</li>
<li>Chromosome variants plots: Plots of number of variants for each chromosome location</li>
<li>Details by gene: Link to "Genes Statistics" file</li>
</ul>
<h2 id="main-summary-table">Main summary table</h2>
<p>The main summary table contains basic information about the SnpEff run and some overall statistics:</p>
<table>
<thead>
<tr>
<th>Table Entry</th>
<th>Note</th>
</tr>
</thead>
<tbody>
<tr>
<td>Genome</td>
<td>Genome name and version, as specified in the command line</td>
</tr>
<tr>
<td>Date</td>
<td>Date and time when the analysis was performed</td>
</tr>
<tr>
<td>SnpEff version</td>
<td>SnpEff version</td>
</tr>
<tr>
<td>Command line arguments</td>
<td>Command line arguments and options used to annotate</td>
</tr>
<tr>
<td>Warnings</td>
<td>Number of <a href="../inputoutput/#errors-and-warnings">WARNING</a> annotation messages (i.e. <code>WARNING</code> messages in the ANN field)</td>
</tr>
<tr>
<td>Errors</td>
<td>Number of <a href="../inputoutput/#errors-and-warnings">ERROR</a> annotation messages (i.e. <code>ERROR</code> messages in the ANN field)</td>
</tr>
<tr>
<td>Number of lines (input file)</td>
<td>Number of lines in the input file, excluding comment / header lines</td>
</tr>
<tr>
<td>Number of variants (before filter)</td>
<td>Number of variants in the input file. Note that this can differ from the number of lines; e.g. VCF allows for <a href="../faq/#counting-variants-vs-lines-in-a-vcf-file">multiple variants per line</a> and <a href="./faq#iupac-expansion">IUPAC expansion</a>.</td>
</tr>
<tr>
<td>Number of not variants</td>
<td>Number of non-variants, e.g. if the <code>REF</code> and <code>ALT</code> fields are the same</td>
</tr>
<tr>
<td>Number of variants processed</td>
<td>Number of variants processed. This can be different than the number of variants due to filtering and non-variants entries.</td>
</tr>
<tr>
<td>Number of known variants</td>
<td>Variants that have a non-empty <code>ID</code> field.</td>
</tr>
<tr>
<td>Number of multi-allelic VCF entries</td>
<td>Variants that have mode then two alleles. Most variants have only two alleles: <code>REF</code> and one <code>ALT</code>. Multi-allelic variants have multiple <code>ALT</code> entries.</td>
</tr>
<tr>
<td>Number of annotations</td>
<td>Total number of variant annotations. Note that this is typically <a href="../faq/#counting-variants-vs-annotations">higher than the number of variant</a></td>
</tr>
<tr>
<td>Genome total length</td>
<td>Total genome length (in bases)</td>
</tr>
<tr>
<td>Genome effective length</td>
<td>Total length of the chromosomes (in bases). This only counts chromosomes that had varinats</td>
</tr>
<tr>
<td>Variant rate</td>
<td>Number of variants per genomic length: <code>Number of variants</code> / <code>Genome effeective length</code></td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The number of input lines, number of variants, and number of annotation are different counts and typically are not equal, <a href="../faq/#number-of-variants-in-vcf-and-html-summary-do-not-match">see details in this FAQ</a></p>
</div>
<h3 id="variants-by-type">Variants by type</h3>
<p>This table contains a list of the number of variants, grouped by variant type:</p>
<table>
<thead>
<tr>
<th>Type</th>
<th>Note</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>SNP</code></td>
<td>SNP / SNV is a single nucleotide variant, e.g. 'A -&gt; G'</td>
</tr>
<tr>
<td><code>MNP</code></td>
<td>MNP / MNV is a multiple nucleotide variant, e.g. 'AC -&gt; GT'</td>
</tr>
<tr>
<td><code>INS</code></td>
<td>Insertion, e.g. 'A -&gt; AT'</td>
</tr>
<tr>
<td><code>DEL</code></td>
<td>Deletion, e.g. 'AT -&gt; A'</td>
</tr>
<tr>
<td><code>MIXED</code></td>
<td>A mixed vairant is a combination of SNP / MNP / INS / DEL, for example '</td>
</tr>
<tr>
<td><code>INV</code></td>
<td>An inversion of reference sequence</td>
</tr>
<tr>
<td><code>DUP</code></td>
<td>A duplication is a region of elevated copy number relative to the reference</td>
</tr>
<tr>
<td><code>BND</code></td>
<td>An arbitrary rearrangement</td>
</tr>
<tr>
<td><code>INTERVAL</code></td>
<td>An interval marke, e.g. an interval from a BED file</td>
</tr>
</tbody>
</table>
<h3 id="histograms">Histograms</h3>
<p>E.g.: In the stats file, you can see coverage histogram plots like this one:</p>
<p><img alt="" src="../../images/coverage_hitogram.png" /></p>
<h3 id="annotations-region">Annotations &amp; Region</h3>
<p>SnpEff annotates variants using <a href="../inputoutput/#variant-annotaiton-details">"functional annotaions"</a>, e.g. <code>NON_SYNONYMOUS_CODING</code>, <code>STOP_GAINED</code>, etc..
These variants affect regions of the genome (e.g. <code>EXON</code>, <code>INTRON</code>).
The two tables count how many effects for each type and for each region exists.</p>
<p>E.g.: In an <code>EXON</code> region, you can have all the following effect types: <code>NON_SYNONYMOUS_CODING</code>, <code>SYNONYMOUS_CODING</code>, <code>FRAME_SHIFT</code>, <code>STOP_GAINED</code>, etc.</p>
<p>The complicated part is that some annotaitons affect a region that has the same name (yes, I know, this is confusing).</p>
<p>E.g.: In a <code>UTR_5_PRIME</code> region you can have <code>UTR_5_PRIME</code> and <code>START_GAINED</code> effect type.</p>
<p>This means that the number of both tables are not exactly the same, because the labels don't mean the same.
See the next figure as an example:</p>
<p><img alt="type_vs_region" class="center" src="../../images/type_vs_region.png" /></p>
<p>So the number of effects that affect a UTR_5_PRIME <em>region</em> is 206. Of those, 57 are effects <em>type</em> <code>START_GAINED</code> and 149 are effects <em>type</em> <code>UTR_5_PRIME</code>.</p>
<p>How exactly are effect <em>type</em> and effect <em>region</em> related? See the following table:</p>
<table>
<thead>
<tr>
<th>Effect Type</th>
<th>Region</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>NONE</code>, <code>CHROMOSOME</code>, <code>CUSTOM</code>, <code>CDS</code></td>
<td><code>NONE</code></td>
</tr>
<tr>
<td><code>INTERGENIC</code>, <code>INTERGENIC_CONSERVED</code></td>
<td><code>INTERGENIC</code></td>
</tr>
<tr>
<td><code>UPSTREAM</code></td>
<td><code>UPSTREAM</code></td>
</tr>
<tr>
<td><code>UTR_5_PRIME</code>, <code>UTR_5_DELETED</code>, <code>START_GAINED</code></td>
<td><code>UTR_5_PRIME</code></td>
</tr>
<tr>
<td><code>SPLICE_SITE_ACCEPTOR</code></td>
<td><code>SPLICE_SITE_ACCEPTOR</code></td>
</tr>
<tr>
<td><code>SPLICE_SITE_DONOR</code></td>
<td><code>SPLICE_SITE_DONOR</code></td>
</tr>
<tr>
<td><code>SPLICE_SITE_REGION</code></td>
<td><code>SPLICE_SITE_REGION</code></td>
</tr>
<tr>
<td><code>INTRAGENIC</code>, <code>START_LOST</code>, <code>SYNONYMOUS_START</code>, <code>NON_SYNONYMOUS_START</code>, <code>GENE</code>, <code>TRANSCRIPT</code></td>
<td><code>EXON</code> or <code>NONE</code></td>
</tr>
<tr>
<td><code>EXON</code>, <code>EXON_DELETED</code>, <code>NON_SYNONYMOUS_CODING</code>, <code>SYNONYMOUS_CODING</code>, <code>FRAME_SHIFT</code>, <code>CODON_CHANGE</code>, <code>CODON_INSERTION</code>, <code>CODON_CHANGE_PLUS_CODON_INSERTION</code>, <code>CODON_DELETION</code>, <code>CODON_CHANGE_PLUS_CODON_DELETION</code>, <code>STOP_GAINED</code>, <code>SYNONYMOUS_STOP</code>, <code>STOP_LOST</code>, <code>RARE_AMINO_ACID</code></td>
<td><code>EXON</code></td>
</tr>
<tr>
<td><code>INTRON</code>, <code>INTRON_CONSERVED</code></td>
<td><code>INTRON</code></td>
</tr>
<tr>
<td><code>UTR_3_PRIME</code>, <code>UTR_3_DELETED</code></td>
<td><code>UTR_3_PRIME</code></td>
</tr>
<tr>
<td><code>DOWNSTREAM</code></td>
<td><code>DOWNSTREAM</code></td>
</tr>
<tr>
<td><code>REGULATION</code></td>
<td><code>REGULATION</code></td>
</tr>
</tbody>
</table>
<h3 id="gene-statistics">Gene statistics</h3>
<p>SnpEff also generates a TXT (tab separated) file having counts of number of variants affecting each transcript and gene.
By default, the file name is <code>snpEff_genes.txt</code>, but it can be changed using the <code>-stats</code> command line option.</p>
<p>Here is an example of this file:
<div class="highlight"><pre><span></span><code>$ head snpEff_genes.txt
# The following table is formatted as tab separated values.
#GeneName   GeneId  TranscriptId    BioType variants_impact_HIGH    variants_impact_LOW variants_impact_MODERATE    variants_impact_MODIFIER    variants_effect_3_prime_UTR_variant variants_effect_5_prime_UTR_premature_start_codon_gain_variant  variants_effect_5_prime_UTR_variant variants_effect_downstream_gene_variant variants_effect_intron_variant  variants_effect_missense_variant    variants_effect_non_coding_exon_variant variants_effect_splice_acceptor_variant variants_effect_splice_donor_variant    variants_effect_splice_region_variant   variants_effect_start_lost  variants_effect_stop_gained variants_effect_stop_lost   variants_effect_synonymous_variant  variants_effect_upstream_gene_variant   bases_affected_DOWNSTREAM   total_score_DOWNSTREAM  length_DOWNSTREAM   bases_affected_EXON total_score_EXON    length_EXON bases_affected_INTRON   total_score_INTRON  length_INTRON   bases_affected_SPLICE_SITE_ACCEPTOR total_score_SPLICE_SITE_ACCEPTOR    length_SPLICE_SITE_ACCEPTOR bases_affected_SPLICE_SITE_DONOR    total_score_SPLICE_SITE_DONOR   length_SPLICE_SITE_DONOR    bases_affected_SPLICE_SITE_REGION   total_score_SPLICE_SITE_REGION  length_SPLICE_SITE_REGION   bases_affected_TRANSCRIPT   total_score_TRANSCRIPT  length_TRANSCRIPT   bases_affected_UPSTREAM total_score_UPSTREAM    length_UPSTREAM bases_affected_UTR_3_PRIME  total_score_UTR_3_PRIME length_UTR_3_PRIME  bases_affected_UTR_5_PRIME  total_score_UTR_5_PRIME length_UTR_5_PRIME
AC000029.1  ENSG00000221069 ENST00000408142 miRNA   0   0   0   2   0   0   0   2   0   0   0   0   0   0   0   0   5000    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
AC000068.5  ENSG00000185065 ENST00000431090 antisense   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   5000    0   0   0   0   0   0
AC000081.2  ENSG00000230194 ENST00000433141 processed_pseudogene    0   0   0   8   0   0   0   3   0   0   0   0   0   0   5000    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   5   0   5000    0   0
AC000089.3  ENSG00000235776 ENST00000424559 processed_pseudogene    0   0   0   1   0   0   0   0   0   0   0   0   0   0   5000    0   0   0   0   0   0
AC002472.1  ENSG00000269103 ENST00000547793 protein_coding  0   0   0   6   0   0   0   5   0   0   0   0   0   0   0   5000    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   1   0   5000    0   0
AC002472.11 ENSG00000226872 ENST00000450652 antisense   0   0   0   13  0   0   0   5   2   0   0   0   0   0   0   5000    0   0   0   2   0   11199   0   0   0   0   0   0   0   0   0   0   0   0   6   0   5000    0   0
AC002472.13 ENSG00000187905 ENST00000342608 protein_coding  0   1   6   1   0   0   0   0   1   6   0   0   0   1   0   116 1   0   934 0   0   0   0   0   0   1   0   3   0   0   0   0   0   0   0   0   0   0   0
AC002472.13 ENSG00000187905 ENST00000442047 protein_coding  0   1   6   1   0   0   0   0   1   6   0   0   0   1   0   116 1   0   934 0   0   0   0   0   0   1   0   3   0   0   0   0   0   0   0   0   0   0   0
</code></pre></div></p>
<p>The columns in this table are:</p>
<table>
<thead>
<tr>
<th>Column name</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>GeneName</td>
<td>Gene name (usually HUGO)</td>
</tr>
<tr>
<td>GeneId</td>
<td>Gene's ID</td>
</tr>
<tr>
<td>TranscriptId</td>
<td>Transcript's ID</td>
</tr>
<tr>
<td>BioType</td>
<td>Transcript's bio-type (if available)</td>
</tr>
<tr>
<td>&nbsp;</td>
<td><strong>The following column is repeated for each impact {HIGH, MODERATE, LOW, MODIFIER}</strong></td>
</tr>
<tr>
<td>variants_impact_*</td>
<td>Count number of variants for each impact category</td>
</tr>
<tr>
<td>&nbsp;</td>
<td><strong>The following column is repeated for each annotated effect (e.g. missense_variant, synonymous_variant, stop_lost, etc.)</strong></td>
</tr>
<tr>
<td>variants_effect_*</td>
<td>Count number of variants for each effect type</td>
</tr>
<tr>
<td>&nbsp;</td>
<td><strong>The following columns are repeated for several genomic regions (DOWNSTREAM, EXON, INTRON, UPSTREAM, etc.)</strong></td>
</tr>
<tr>
<td>bases_affected_*</td>
<td>Number of bases that variants overlap genomic region</td>
</tr>
<tr>
<td>total_score_*</td>
<td>Sum of scores overlapping this genomic region. Note: Scores are only available when input files are type 'BED' (e.g. when annotating ChipSeq experiments)</td>
</tr>
<tr>
<td>length_*</td>
<td>Genomic region length</td>
</tr>
</tbody>
</table>





                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Pablo Cingolani
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="http://www.linkedin.com/in/pablocingolani" target="_blank" rel="noopener" title="www.linkedin.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"/></svg>
    </a>
  
    
    
    
    
      
      
    
    <a href="https://github.com/pcingola/SnpEff" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 480 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "search.highlight", "search.suggest", "toc.integrate", "content.tabs.link", "content.code.annotate", "content.code.copy"], "search": "../../assets/javascripts/workers/search.74e28a9f.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.220ee61c.min.js"></script>
      
    
  </body>
</html>